File: server/app/services/diagram_synthesizer.py
"""
CogniLens Domain-Aware Diagram Synthesizer
Generates topic-aligned, educational Mermaid.js diagrams with concrete information in each node:
- Flowcharts (Process & Logic Flow)
- Mind Maps (Concept Taxonomy & Hierarchy)
- Sequence Diagrams (Timeline & Actor Interactions)
- State Diagrams (Lifecycles & Transitions)
- Class Diagrams (Data Structures & Models)

ZERO generic boilerplate: Every node and explanation is dynamically generated and grounded
in the user's topic and provided text, featuring vivid real-world analogies in plain English.
"""

import re
from typing import Dict, List, Optional, Tuple


def sanitize_mermaid_label(text: str) -> str:
    """Clean label text so it doesn't break Mermaid syntax."""
    cleaned = text.replace('"', "'").replace("\n", " ").replace("(", "（").replace(")", "）").strip()
    return cleaned[:65] if len(cleaned) > 65 else cleaned


def sanitize_mindmap_text(text: str) -> str:
    """Clean mindmap node text (Mermaid mindmap breaks on (), [], {}, and quotes)."""
    cleaned = re.sub(r'[\(\)\[\]\{\}"\':]', "", text).strip()
    return cleaned[:50] if len(cleaned) > 50 else cleaned


def extract_concept_elements(topic: str, text: str) -> Dict[str, List[str]]:
    """
    Extract meaningful technical concepts, actions, entities, and conditions
    from source text and topic to build custom, non-generic diagrams.
    """
    combined = f"{topic}. {text}" if text else topic
    
    # 1. Clean and split sentences
    raw_sentences = [
        s.strip() for s in re.split(r'[.\n;•*]', combined)
        if len(s.strip()) >= 10 and not s.strip().startswith("http")
    ]
    
    # 2. Extract bullet points / lines
    lines = [
        re.sub(r'^[- *•0-9.)]+', '', line).strip()
        for line in combined.splitlines()
        if len(line.strip()) >= 8
    ]
    
    # 3. Extract key noun-like phrases or technical terms
    words = re.findall(r'\b[A-Za-z0-9_-]{3,25}\b', combined)
    stop_words = {
        "this", "that", "with", "from", "have", "were", "what", "which",
        "there", "their", "about", "could", "would", "these", "other",
        "into", "more", "also", "some", "time", "than", "them", "very", "when"
    }
    key_terms = []
    for w in words:
        wl = w.lower()
        if wl not in stop_words and len(wl) > 3 and wl not in [k.lower() for k in key_terms]:
            key_terms.append(w)
            if len(key_terms) >= 12:
                break

    # 4. Extract action phrases (sentences with active verbs)
    action_phrases = []
    for s in raw_sentences:
        clean_s = re.sub(r'^[0-9]+[.)]\s*', '', s).strip()
        if 15 <= len(clean_s) <= 80:
            action_phrases.append(clean_s)
        if len(action_phrases) >= 8:
            break

    # 5. Extract conditional / decision statements
    condition_phrases = []
    for s in raw_sentences:
        lower_s = s.lower()
        if any(w in lower_s for w in ["if ", "when ", "check", "verify", "is ", "whether", "ensures", "unless"]):
            condition_phrases.append(s.strip())
        if len(condition_phrases) >= 4:
            break

    return {
        "sentences": raw_sentences,
        "actions": action_phrases,
        "conditions": condition_phrases,
        "terms": key_terms,
        "lines": lines,
    }


def generate_layman_analogy(topic: str, text: str) -> str:
    """Generate a relatable, non-generic real-world everyday analogy tailored to the topic."""
    t_lower = topic.lower()
    c_lower = (topic + " " + text).lower()

    if any(k in c_lower for k in ["deadlock", "mutex", "lock", "concurrency", "race condition"]):
        return (
            f"Think of {topic} like a busy 4-way intersection where four cars arrive simultaneously. "
            "Each driver waits for the car on their right to go first, so everyone is stuck bumper-to-bumper. "
            "No one can move forward until a traffic cop steps in and tells one car to back up and yield."
        )
    elif any(k in c_lower for k in ["schedule", "quantum", "round robin", "cpu", "fifo", "priority"]):
        return (
            f"Think of {topic} like a fair playground where only one swing is available for a line of eager kids. "
            "Instead of letting one kid stay on the swing all afternoon, a supervisor sets a 2-minute timer. "
            "When the buzzer sounds, the current kid steps off to join the back of the queue, and the next kid gets their turn."
        )
    elif any(k in c_lower for k in ["paging", "virtual memory", "cache", "tlb", "ram"]):
        return (
            f"Think of {topic} like working on a research paper at a compact study desk with a vast library downstairs. "
            "Your desk (RAM) only holds 3 reference books at a time. When you need facts from a 4th book, "
            "you swap one book back to the basement shelf (Disk) and retrieve the new book onto your desk."
        )
    elif any(k in c_lower for k in ["tcp", "handshake", "packet", "socket", "network", "protocol"]):
        return (
            f"Think of {topic} like an important phone call with a spotty connection. "
            "Before sharing private details, you say 'Can you hear me clearly?', the receiver replies 'Yes, I hear you, can you hear me?', "
            "and you confirm 'Got you, let's talk!'. Only once both parties verify the link does the actual conversation begin."
        )
    elif any(k in c_lower for k in ["tree", "bst", "binary search", "graph", "dijkstra", "sort", "algorithm"]):
        return (
            f"Think of {topic} like organizing a massive dictionary or phone book. "
            "Rather than flipping through every single page from A to Z, you open directly to the middle, check if your target word comes before or after, "
            "and instantly throw away half of the book at every step until you pinpoint the exact word in seconds."
        )
    elif any(k in c_lower for k in ["neural", "backprop", "gradient", "loss", "deep learning", "machine learning"]):
        return (
            f"Think of {topic} like learning to shoot basketball free throws while wearing fogged-up glasses. "
            "After every missed shot, your coach tells you whether the ball was two inches too far to the left or short. "
            "You nudge your wrist angle slightly in the opposite direction on the next shot until the ball swishes through the net consistently."
        )
    elif any(k in c_lower for k in ["database", "transaction", "acid", "sql", "commit", "rollback"]):
        return (
            f"Think of {topic} like transferring money between bank accounts at an ATM. "
            "Either the money is debited from your checking account AND credited to your savings, or neither happens. "
            "If the ATM loses power mid-transfer, it undoes the half-finished action completely so your money is never lost."
        )
    else:
        return (
            f"Think of {topic} like an assembly line recipe in a bakery: "
            "every ingredient and stage relies on the previous step being verified first. If any measurement is off, "
            "the system detects the discrepancy early and adjusts before finalizing the finished product."
        )


def generate_synthesized_diagram(
    topic: str,
    text_content: str,
    diagram_type: str = "flowchart",
    direction: str = "TD",
) -> Tuple[str, str, str, str, List[str]]:
    """
    Generate an educational, non-generic Mermaid.js diagram and simplified explanation
    grounded directly in the provided topic and text content.
    Returns: (mermaid_code, title, description, simplified_explanation, key_takeaways)
    """
    clean_topic = topic.strip() or "Core Subject"
    safe_topic = sanitize_mermaid_label(clean_topic)
    elements = extract_concept_elements(clean_topic, text_content)
    
    actions = elements["actions"]
    conditions = elements["conditions"]
    terms = elements["terms"]
    
    layman_analogy = generate_layman_analogy(clean_topic, text_content)

    # -------------------------------------------------------------------------
    # 1. MIND MAP
    # -------------------------------------------------------------------------
    if diagram_type == "mindmap":
        title = f"Mind Map: {clean_topic}"
        description = f"Taxonomy and core mechanisms of {clean_topic}."
        root = sanitize_mindmap_text(clean_topic)
        
        # Build 4 topic-grounded branches
        b1 = sanitize_mindmap_text(actions[0]) if len(actions) > 0 else f"{root} Foundations"
        b2 = sanitize_mindmap_text(actions[1]) if len(actions) > 1 else "Core Execution Mechanism"
        b3 = sanitize_mindmap_text(actions[2]) if len(actions) > 2 else "Key Constraints & Rules"
        b4 = sanitize_mindmap_text(actions[3]) if len(actions) > 3 else "Practical Applied Impact"
        
        t1 = sanitize_mindmap_text(terms[0]) if len(terms) > 0 else "Primary Component"
        t2 = sanitize_mindmap_text(terms[1]) if len(terms) > 1 else "State Validation"
        t3 = sanitize_mindmap_text(terms[2]) if len(terms) > 2 else "Invariant Rule"
        t4 = sanitize_mindmap_text(terms[3]) if len(terms) > 3 else "Verified Outcome"

        code = f"""mindmap
  root(( {root} ))
    Fundamental Concepts
      {b1}
      {t1}
    Operational Flow
      {b2}
      {t2}
    Constraints & Conditions
      {b3}
      {t3}
    Applied Results
      {b4}
      {t4}"""

        takeaways = [
            f"Core principle: {b1[:60]}.",
            f"Operational flow: {b2[:60]}.",
            f"Key takeaway: {b4[:60]}."
        ]
        return code.strip(), title, description, layman_analogy, takeaways

    # -------------------------------------------------------------------------
    # 2. SEQUENCE DIAGRAM
    # -------------------------------------------------------------------------
    elif diagram_type == "sequence":
        title = f"Sequence Flow: {clean_topic}"
        description = f"Step-by-step actor interactions and timeline of {clean_topic}."
        
        actor1 = sanitize_mermaid_label(terms[0] if len(terms) > 0 else "Client").replace(" ", "")[:18]
        actor2 = sanitize_mermaid_label(terms[1] if len(terms) > 1 else "Coordinator").replace(" ", "")[:18]
        actor3 = sanitize_mermaid_label(terms[2] if len(terms) > 2 else "Engine").replace(" ", "")[:18]
        
        step1 = sanitize_mermaid_label(actions[0] if len(actions) > 0 else f"Initiate {clean_topic}")[:45]
        step2 = sanitize_mermaid_label(actions[1] if len(actions) > 1 else "Validate preconditions and parameters")[:45]
        step3 = sanitize_mermaid_label(actions[2] if len(actions) > 2 else "Process core operation")[:45]
        step4 = sanitize_mermaid_label(actions[3] if len(actions) > 3 else "Return verified response")[:45]
        cond = sanitize_mermaid_label(conditions[0] if conditions else "Condition Satisfied")[:35]

        code = f"""sequenceDiagram
    autonumber
    actor A as {actor1}
    participant B as {actor2}
    participant C as {actor3}

    A->>B: {step1}
    B->>C: {step2}
    C-->>B: Status Check ({cond})
    alt {cond}
        B->>C: Execute: {step3}
        C-->>B: Execution Success
        B-->>A: {step4}
    else Constraint Violation
        C-->>B: Emit Warning / Retry
        B-->>A: Safe Fallback / Adjusted Signal
    end"""

        takeaways = [
            f"{actor1} initiates operation with {step1[:50]}.",
            f"{actor2} coordinates verification against {cond[:40]}.",
            f"Terminal state produces verified outcome {step4[:50]}."
        ]
        return code.strip(), title, description, layman_analogy, takeaways

    # -------------------------------------------------------------------------
    # 3. STATE DIAGRAM
    # -------------------------------------------------------------------------
    elif diagram_type == "stateDiagram":
        title = f"State Machine: {clean_topic}"
        description = f"Lifecycle transitions and triggers for {clean_topic}."
        
        s1 = sanitize_mermaid_label(terms[0] if len(terms) > 0 else "Idle").replace(" ", "")[:16]
        s2 = sanitize_mermaid_label(terms[1] if len(terms) > 1 else "Processing").replace(" ", "")[:16]
        s3 = sanitize_mermaid_label(terms[2] if len(terms) > 2 else "Verified").replace(" ", "")[:16]
        
        t1 = sanitize_mermaid_label(actions[0] if len(actions) > 0 else "Start Request")[:30]
        t2 = sanitize_mermaid_label(actions[1] if len(actions) > 1 else "Core Computation")[:30]
        t3 = sanitize_mermaid_label(actions[2] if len(actions) > 2 else "Validation Passed")[:30]

        code = f"""stateDiagram-v2
    [*] --> {s1} : System Initialized
    {s1} --> {s2} : {t1}
    {s2} --> {s3} : {t2}
    {s3} --> SuccessComplete : {t3}
    {s2} --> ErrorRecovery : Failure / Boundary Trigger
    ErrorRecovery --> {s1} : Reset to Safe State
    SuccessComplete --> [*] : Lifecycle Completed"""

        takeaways = [
            f"Enters {s1} upon initialization.",
            f"Transitions to {s2} upon {t1}.",
            f"Achieves terminal completion via {t3}."
        ]
        return code.strip(), title, description, layman_analogy, takeaways

    # -------------------------------------------------------------------------
    # 4. CLASS DIAGRAM
    # -------------------------------------------------------------------------
    elif diagram_type == "class":
        title = f"Class Model: {clean_topic}"
        description = f"Structural representation and methods of {clean_topic}."
        
        c1 = sanitize_mermaid_label(terms[0] if len(terms) > 0 else "PrimaryEntity").replace(" ", "")[:18]
        c2 = sanitize_mermaid_label(terms[1] if len(terms) > 1 else "Manager").replace(" ", "")[:18]
        c3 = sanitize_mermaid_label(terms[2] if len(terms) > 2 else "Handler").replace(" ", "")[:18]
        
        m1 = sanitize_mermaid_label(actions[0] if len(actions) > 0 else "executeStep")[:22].replace(" ", "")
        m2 = sanitize_mermaid_label(actions[1] if len(actions) > 1 else "validateState")[:22].replace(" ", "")

        code = f"""classDiagram
    class {c1} {{
        +String identifier
        +Boolean isActive
        +{m1}() void
    }}
    class {c2} {{
        -List items
        +process() bool
        +{m2}() bool
    }}
    class {c3} {{
        +notify() void
        +handleResult() void
    }}
    {c1} "1" --> "many" {c2} : coordinates
    {c2} ..> {c3} : outputs to"""

        takeaways = [
            f"{c1} acts as the root entity managing configuration.",
            f"{c2} handles execution and invokes {m2}().",
            f"{c3} captures output and boundary states."
        ]
        return code.strip(), title, description, layman_analogy, takeaways

    # -------------------------------------------------------------------------
    # 5. FLOWCHART (Default)
    # -------------------------------------------------------------------------
    else:
        title = f"Flowchart: {clean_topic}"
        description = f"Step-by-step logic, conditions, and outcomes of {clean_topic}."
        dir_code = "LR" if direction == "LR" else "TD"
        
        step1 = sanitize_mermaid_label(actions[0] if len(actions) > 0 else f"Initialize {clean_topic}")
        step2 = sanitize_mermaid_label(actions[1] if len(actions) > 1 else "Evaluate core parameters & inputs")
        
        cond = sanitize_mermaid_label(
            conditions[0] if len(conditions) > 0 else (actions[2] if len(actions) > 2 else "Are requirements satisfied?")
        )
        if not cond.endswith("?"):
            cond = cond[:40] + "?"
            
        step3 = sanitize_mermaid_label(actions[2] if len(actions) > 2 else "Apply primary transformation")
        step4 = sanitize_mermaid_label(actions[3] if len(actions) > 3 else "Generate final outcome")

        code = f"""flowchart {dir_code}
    A["Input / Trigger: {safe_topic}"] --> B["{step1}"]
    B --> C["{step2}"]
    C --> D{{"{cond}"}}
    D -->|Yes: Valid| E["{step3}"]
    D -->|No: Alternate| F["Apply fallback handling & recovery"]
    F --> E
    E --> G["Result: Successfully completed {safe_topic[:30]}"]

    style A fill:#2563eb,stroke:#1d4ed8,color:#fff
    style C fill:#7c3aed,stroke:#6d28d9,color:#fff
    style D fill:#d97706,stroke:#b45309,color:#fff
    style G fill:#059669,stroke:#047857,color:#fff"""

        takeaways = [
            f"Step 1: {step1[:55]}.",
            f"Condition check: {cond[:55]}.",
            f"Target outcome: {step4[:55]}."
        ]
        return code.strip(), title, description, layman_analogy, takeaways

File: client/src/services/diagramSynthesizer.ts
/**
 * CogniLens Domain-Aware Client Diagram Synthesizer
 * Generates topic-aligned, non-generic Mermaid.js diagrams for:
 * - Flowchart (Process & Flow)
 * - Mind Map (Concept Taxonomy & Hierarchy)
 * - Sequence Diagram (Timeline & Actor Interactions)
 * - State Diagram (Lifecycles & Transitions)
 * - Class Diagram (Data Structures & Models)
 *
 * ZERO generic boilerplate: extracts genuine concepts, terms, actions, and conditions
 * from the user's provided topic and text, accompanied by rich plain-English layman analogies.
 */

export interface SynthesizedDiagramResult {
  mermaid_code: string;
  title: string;
  description: string;
  simplified_explanation?: string;
  key_takeaways?: string[];
}

export function sanitizeMermaidLabel(text: string): string {
  const cleaned = text.replace(/"/g, "'").replace(/\n/g, ' ').replace(/\(/g, '（').replace(/\)/g, '）').trim();
  return cleaned.length > 65 ? cleaned.slice(0, 65) : cleaned;
}

export function sanitizeMindmapText(text: string): string {
  const cleaned = text.replace(/[()[\]{}"':]/g, '').trim();
  return cleaned.length > 48 ? cleaned.slice(0, 48) : cleaned;
}

export function extractConceptElements(topic: string, text: string): {
  actions: string[];
  conditions: string[];
  terms: string[];
} {
  const combined = text ? `${topic}. ${text}` : topic;

  // Extract sentence fragments
  const sentences = combined
    .split(/[.\n;•*]/)
    .map(s => s.replace(/^[0-9]+[.)]\s*/, '').trim())
    .filter(s => s.length >= 10 && !s.startsWith('http'));

  const actionPhrases = sentences.filter(s => s.length >= 12 && s.length <= 80).slice(0, 8);

  const conditionPhrases = sentences.filter(s => {
    const l = s.toLowerCase();
    return l.includes('if ') || l.includes('when ') || l.includes('verify') || l.includes('check') || l.includes('ensure') || l.includes('is ');
  }).slice(0, 4);

  // Key terms
  const words = combined.match(/\b[A-Za-z0-9_-]{3,25}\b/g) || [];
  const stopWords = new Set([
    'this', 'that', 'with', 'from', 'have', 'were', 'what', 'which',
    'there', 'their', 'about', 'could', 'would', 'these', 'other',
    'into', 'more', 'also', 'some', 'time', 'than', 'them', 'very', 'when'
  ]);
  const terms: string[] = [];
  for (const w of words) {
    const wl = w.toLowerCase();
    if (!stopWords.has(wl) && wl.length > 3 && !terms.some(t => t.toLowerCase() === wl)) {
      terms.push(w);
      if (terms.length >= 10) break;
    }
  }

  return {
    actions: actionPhrases,
    conditions: conditionPhrases,
    terms
  };
}

export function generateLaymanAnalogy(topic: string, text: string): string {
  const combined = `${topic} ${text}`.toLowerCase();

  if (/(deadlock|mutex|lock|concurrency|race condition)/.test(combined)) {
    return `Think of ${topic} like a busy 4-way traffic intersection where four cars arrive at the same instant. Each driver waits for the car on their right to proceed, resulting in a complete standstill where nobody can move. The gridlock only breaks when a traffic officer steps in and commands one vehicle to back up.`;
  } else if (/(schedule|quantum|round robin|cpu|fifo|priority)/.test(combined)) {
    return `Think of ${topic} like a fair playground where a dozen kids want to use a single swing. Instead of letting one child hog the swing for an hour, a timer is set for 2 minutes. When the buzzer goes off, the current kid hops off to the back of the line, ensuring everyone gets regular turns with predictable waiting times.`;
  } else if (/(paging|virtual memory|cache|tlb|ram)/.test(combined)) {
    return `Think of ${topic} like working on a major research report at a small desk with an enormous archive in the cellar. Your desk (RAM) only holds three open books at once. When you need information from a fourth book, you return one book to the cellar shelves and bring the new volume up to your desk.`;
  } else if (/(tcp|handshake|packet|socket|network|protocol)/.test(combined)) {
    return `Think of ${topic} like a formal radio check before a space launch. You broadcast 'Control, do you read me?' (SYN). Control answers 'Loud and clear, do you copy us?' (SYN-ACK). You reply 'Copy loud and clear, ready for telemetry' (ACK). Only after this mutual confirmation is real mission data transmitted.`;
  } else if (/(tree|bst|binary search|graph|dijkstra|sort|algorithm)/.test(combined)) {
    return `Think of ${topic} like searching for a word in a dictionary. Rather than reading every single word from page 1 onwards, you split the dictionary in half, see which half contains your letter, and immediately discard the other half, finding any entry in just a few quick checks.`;
  } else if (/(neural|backprop|gradient|loss|deep learning|machine learning)/.test(combined)) {
    return `Think of ${topic} like a chef seasoning a complex soup recipe. After tasting each batch, if it's slightly too salty, the chef reduces the salt measurement by a tiny pinch next time. Step by step, tasting and tweaking after each batch, the recipe approaches perfection.`;
  } else if (/(database|transaction|acid|sql|commit|rollback)/.test(combined)) {
    return `Think of ${topic} like buying a coffee with a debit card. Either the store receives payment AND you get your coffee, or neither happens. If the payment terminal freezes midway, the transaction aborts cleanly so your bank balance remains untouched.`;
  }

  return `Think of ${topic} like a precision assembly line where each stage builds on the verified output of the previous step. If any component fails quality inspection, the system adjusts or safely falls back before packaging the final result.`;
}

export function generateClientSynthesizedDiagram(
  topic: string,
  textContent: string,
  diagramType: string = 'flowchart',
  direction: string = 'TD'
): SynthesizedDiagramResult {
  const cleanTopic = topic.trim() || 'Core Concept';
  const safeTopic = sanitizeMermaidLabel(cleanTopic);
  const elements = extractConceptElements(cleanTopic, textContent);

  const { actions, conditions, terms } = elements;
  const laymanAnalogy = generateLaymanAnalogy(cleanTopic, textContent);

  // 1. MIND MAP
  if (diagramType === 'mindmap') {
    const root = sanitizeMindmapText(cleanTopic);
    const title = `Mind Map: ${cleanTopic}`;
    const description = `Core concepts and relationships for ${cleanTopic}.`;

    const b1 = sanitizeMindmapText(actions[0] || `${root} Fundamentals`);
    const b2 = sanitizeMindmapText(actions[1] || 'Core Mechanics');
    const b3 = sanitizeMindmapText(actions[2] || 'Governing Rules & Limits');
    const b4 = sanitizeMindmapText(actions[3] || 'Real-World Applications');

    const t1 = sanitizeMindmapText(terms[0] || 'Primary Component');
    const t2 = sanitizeMindmapText(terms[1] || 'State Verification');
    const t3 = sanitizeMindmapText(terms[2] || 'Invariant Property');
    const t4 = sanitizeMindmapText(terms[3] || 'Verified Result');

    const code = `mindmap
  root(( ${root} ))
    Foundational Principles
      ${b1}
      ${t1}
    Operational Workflow
      ${b2}
      ${t2}
    Rules & Boundary Limits
      ${b3}
      ${t3}
    Practical Applications
      ${b4}
      ${t4}`;

    return {
      mermaid_code: code,
      title,
      description,
      simplified_explanation: laymanAnalogy,
      key_takeaways: [
        `Core principle: ${b1.slice(0, 50)}.`,
        `Mechanism: ${b2.slice(0, 50)}.`,
        `Practical application: ${b4.slice(0, 50)}.`
      ]
    };
  }

  // 2. SEQUENCE DIAGRAM
  if (diagramType === 'sequence') {
    const title = `Sequence Flow: ${cleanTopic}`;
    const description = `Timeline and participant interactions for ${cleanTopic}.`;

    const act1 = (terms[0] || 'Client').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);
    const act2 = (terms[1] || 'Coordinator').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);
    const act3 = (terms[2] || 'Engine').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);

    const s1 = sanitizeMermaidLabel(actions[0] || `Initiate ${cleanTopic}`).slice(0, 42);
    const s2 = sanitizeMermaidLabel(actions[1] || 'Validate preconditions').slice(0, 42);
    const s3 = sanitizeMermaidLabel(actions[2] || 'Process core transformation').slice(0, 42);
    const s4 = sanitizeMermaidLabel(actions[3] || 'Return confirmed result').slice(0, 42);
    const cond = sanitizeMermaidLabel(conditions[0] || 'Preconditions Validated').slice(0, 32);

    const code = `sequenceDiagram
    autonumber
    actor A as ${act1}
    participant B as ${act2}
    participant C as ${act3}

    A->>B: ${s1}
    B->>C: ${s2}
    C-->>B: Status Check (${cond})
    alt ${cond}
        B->>C: Execute: ${s3}
        C-->>B: Execution Success
        B-->>A: ${s4}
    else Boundary Check Failed
        C-->>B: Emit Warning / Retry
        B-->>A: Safe Fallback Signal
    end`;

    return {
      mermaid_code: code,
      title,
      description,
      simplified_explanation: laymanAnalogy,
      key_takeaways: [
        `${act1} triggers the workflow with ${s1.slice(0, 45)}.`,
        `${act2} verifies rules against ${cond.slice(0, 35)}.`,
        `Result: ${s4.slice(0, 45)}.`
      ]
    };
  }

  // 3. STATE DIAGRAM
  if (diagramType === 'stateDiagram') {
    const title = `State Machine: ${cleanTopic}`;
    const description = `Lifecycle transitions and event triggers for ${cleanTopic}.`;

    const s1 = (terms[0] || 'InitialState').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);
    const s2 = (terms[1] || 'ActiveProcessing').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);
    const s3 = (terms[2] || 'VerifiedState').replace(/[^a-zA-Z0-9]/g, '').slice(0, 16);

    const t1 = sanitizeMermaidLabel(actions[0] || 'Start Request').slice(0, 28);
    const t2 = sanitizeMermaidLabel(actions[1] || 'Compute Transformation').slice(0, 28);
    const t3 = sanitizeMermaidLabel(actions[2] || 'Invariants Confirmed').slice(0, 28);

    const code = `stateDiagram-v2
    [*] --> ${s1} : Initialize
    ${s1} --> ${s2} : ${t1}
    ${s2} --> ${s3} : ${t2}
    ${s3} --> SuccessCompleted : ${t3}
    ${s2} --> ErrorRecovery : Boundary Failure
    ErrorRecovery --> ${s1} : Reset to Safe Bounds
    SuccessCompleted --> [*] : Lifecycle Completed`;

    return {
      mermaid_code: code,
      title,
      description,
      simplified_explanation: laymanAnalogy,
      key_takeaways: [
        `Starts in ${s1} on initialization.`,
        `Transitions to ${s2} when ${t1.slice(0, 40)}.`,
        `Reaches completion upon ${t3.slice(0, 40)}.`
      ]
    };
  }

  // 4. CLASS DIAGRAM
  if (diagramType === 'class') {
    const title = `Class Model: ${cleanTopic}`;
    const description = `Entities and relationships of ${cleanTopic}.`;

    const c1 = (terms[0] || 'PrimaryModel').replace(/[^a-zA-Z0-9]/g, '').slice(0, 18);
    const c2 = (terms[1] || 'Manager').replace(/[^a-zA-Z0-9]/g, '').slice(0, 18);
    const c3 = (terms[2] || 'OutputHandler').replace(/[^a-zA-Z0-9]/g, '').slice(0, 18);

    const m1 = (actions[0] || 'execute').replace(/[^a-zA-Z0-9]/g, '').slice(0, 18) || 'execute';
    const m2 = (actions[1] || 'validate').replace(/[^a-zA-Z0-9]/g, '').slice(0, 18) || 'validate';

    const code = `classDiagram
    class ${c1} {
        +String identifier
        +Boolean isActive
        +${m1}() void
    }
    class ${c2} {
        -List items
        +process() bool
        +${m2}() bool
    }
    class ${c3} {
        +notify() void
        +handleResult() void
    }
    ${c1} "1" --> "many" ${c2} : coordinates
    ${c2} ..> ${c3} : outputs to`;

    return {
      mermaid_code: code,
      title,
      description,
      simplified_explanation: laymanAnalogy,
      key_takeaways: [
        `${c1} acts as the core entity coordinator.`,
        `${c2} executes operations via ${m2}().`,
        `${c3} encapsulates outputs and notifications.`
      ]
    };
  }

  // 5. FLOWCHART (Default)
  const dirCode = direction === 'LR' ? 'LR' : 'TD';
  const title = `Flowchart: ${cleanTopic}`;
  const description = `Step-by-step logic, conditions, and outcomes of ${cleanTopic}.`;

  const step1 = sanitizeMermaidLabel(actions[0] || `Initialize ${cleanTopic}`).slice(0, 45);
  const step2 = sanitizeMermaidLabel(actions[1] || 'Evaluate core parameters and inputs').slice(0, 45);
  let cond = sanitizeMermaidLabel(conditions[0] || (actions[2] || 'Are constraints satisfied?')).slice(0, 40);
  if (!cond.endsWith('?')) cond = `${cond}?`;
  const step3 = sanitizeMermaidLabel(actions[2] || 'Apply primary transformation').slice(0, 45);
  const step4 = sanitizeMermaidLabel(actions[3] || 'Produce verified outcome').slice(0, 45);

  const code = `flowchart ${dirCode}
    A["Input / Trigger: ${safeTopic.slice(0, 32)}"] --> B["${step1}"]
    B --> C["${step2}"]
    C --> D{"${cond}"}
    D -->|Yes: Valid| E["${step3}"]
    D -->|No: Alternate| F["Apply fallback handling & recovery"]
    F --> E
    E --> G["Result: Completed ${safeTopic.slice(0, 30)}"]

    style A fill:#2563eb,stroke:#1d4ed8,color:#fff
    style C fill:#7c3aed,stroke:#6d28d9,color:#fff
    style D fill:#d97706,stroke:#b45309,color:#fff
    style G fill:#059669,stroke:#047857,color:#fff`;

  return {
    mermaid_code: code,
    title,
    description,
    simplified_explanation: laymanAnalogy,
    key_takeaways: [
      `Step 1: ${step1}.`,
      `Verification: ${cond}.`,
      `Outcome: ${step4}.`
    ]
  };
}
